src/natcap/invest/sdr/sdr_core.pyx

```diff
@@ -154,28 +159,35 @@ def calculate_sediment_deposition(
                     # the weighted sum of flux flowing onto this pixel from
                     # all neighbors
                     f_j_weighted_sum = 0
-                    for neighbor in (
-                            mfd_flow_direction_raster.get_upslope_neighbors(
-                                global_col, global_row)):
+                    up_iterator = UpslopeNeighborIterator(
+                        mfd_flow_direction_raster, global_col, global_row)
+                    neighbor = up_iterator.next()
+                    while neighbor.direction < 8:
```

Comment:
Same question here about whether it'd make sense to track a neighbor/next neighbor object in the loop to make it easier on us to maintain and avoid infinite loops.